fix: Remove `properties`, `options` and `configurators` aggregations if no filterable properties are found

aragon999 opened the pull request:
### 1. Why is this change necessary?
Currently the `properties`, `options` and `configurators` aggregations are only removed if ids are found. This results into the filter panel being rendered even if empty as there still exist aggregations:
https://github.com/shopware/platform/blob/8b1d7518ef1833212a38fe939b87ab98420cbe27/src/Storefront/Resources/views/storefront/element/cms-element-sidebar-filter.html.twig#L10

### 2. What does this change do, exactly?
Always remove the `properties`, `options` and `configurators` aggregations and only add the `properties` aggregation if related `properties` are found.

### 3. Describe each step to reproduce the issue or behaviour.
Create a category, assign one product without any properties, assign a layout which has all filters disabled, except the property filters, the filter panel is still rendered, and on mobile one sees a filter button.

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### 6. Additional commits
I additionally pushed two more commits, one is fixing the phpstan issues of the `ProductListingFeatureSubscriber` and the other is fixing an issue with the tests, where a wrong array shape was passed. 

I noticed this as I looked into the phpstan problems of the test. But I currently have no time to add the array shapes of the data providers :-) Maybe one wants to ignore these altogether, as described here: https://github.com/phpstan/phpstan-phpunit/issues/63#issuecomment-592515871 

<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2650"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

aragon999 committed:
fix: Remove `properties`, `options` and `configurators` aggregations if no filterable properties are found

aragon999 committed:
chore: Fix phpstan issues

SimonVorgers commented:
Hello @aragon999, 
our quality assurance has found out that this behavior also occurs with other filters like 'Rating'. Maybe we can find a general solution that solves the problem for all filters.

aragon999 commented:
@SimonVorgers what exactly do you mean, that the filters are shown if no rating exists for the products of this category?

aragon999 commented:
@SimonVorgers I have looked into this issue and understood the problem.

To solve it each of the other filters needs filtering, but I think the exact filtering depends on the filter itself, for example the rating should be removed if `max` is `null`:
![grafik](https://user-images.githubusercontent.com/6317761/188624153-2d113a3b-a845-4a05-96f9-bc1e9e70f50e.png)

I will look into this further, but not sure when I will be able to fix this.

aragon999 commented:
@SimonVorgers I added a method which filters the empty aggregations. Unfortunately I currently have no time to look into the failing ES test.
The test is failing since the rating aggregation is missing. But not sure how to properly fix that. Could you look into that?

aragon999 commented:
@SimonVorgers I now fixed all tests, rebased to the current trunk and also found another issue which happend in the search, that the filter were still displayed. I therefore extended the filter procedure. 

Could please you review the changes? :-)

aragon999 commented:
@SimonVorgers how is the status here?

aragon999 commented:
@SimonVorgers @keulinho @trandangtri what is the status here? It is quite annoying to have to wait always for such a long time, without _any_ feedback.
There was only one feedback (over a month ago), but since than nothing. By now the phpstan issues of this PR have already been implemented, which now cause a merge conflicts of this PR.

As this is for a customer, I also need to carry around this changes as a patch and apply it to a project. I can now refactor this PR again, to take the phpstan changes of `trunk` into account, but I am not willing to do it a second time.

Edit: Additionally fixing the phpstan issues also costs, in this case my free time, so if this happens more regularly, I wont fix these issues in future (which means more work for you if I see it correctly).

SimonVorgers commented:
Hi @aragon999 
we have re-imported this pull request and are reviewing it with our QA.
Also we are taking care of some additional things like conflicts / fixing some phpstan issues that occur in our gitlab pipeline :)

aragon999 commented:
@SimonVorgers thank you :-)

SimonVorgers commented:
Hi @aragon999 
sorry for writing so late (again). Your PR was already in the merge train, but failed. We noticed that your fix is not compatible with the option `Disable filter options without results` under product settings. I had already looked at the problem but didn`t find a nice solution for it. :disappointed:

aragon999 commented:
To be honest, this PR becomes really annoying. In particular since I do not know how often I rebased, than some things have already been implemented, which this PR included, e.g., the fix for the tests: https://github.com/shopware/platform/commit/b43f8877752b3950ce6aad6c22164783386514ce#diff-05caeea29d8f9461d5aeba86e15cf9714c2c6b728deae62d464804397572f5a4

In future it would be for example nice, if you cannot merge the full PR, that you cherry pick the parts, which can be merged... In this case it could have been the fix for the tests, or the PHPStan issues. But otherwise rebasing and than noticing that it has already been fixed otherwise is annoying... I know this is probably some more work on your side, but otherwise being a contributor is not fun. And in the meantime you on your side have to again invest the time to fix an already fixed issue (if you do not know that this fix exists somewhere).
If for example some parts should than be put into a separate PR, fine, I can do that, just let me know... -.-
Taking so much time between answers is also not helping, as the core changes a lot, I mean the PR was opened in August and will now only land significant time later, probably only in the next major release.

Now coming back to the listing, after rebasing and cleaning up, I tried to reproduce the "issue". But to be honest I have no idea what you mean. For me everything works, as intended, i.e. with the configuration "Disable filter options without results" enabled, for example the "shoe-size" selection is disabled if only filters are selected, which do not have a corresponding "shoe-size":
![grafik](https://user-images.githubusercontent.com/6317761/212547712-243a6f9a-5e2a-49da-9994-910f0ae0914e.png)

So could you please elaborate a bit more on this, what exactly does not work in which configuration?

Edit: I pushed my rebased version of this, it would be good, if you could rebase your changes as well on mine, and push them, as the issue might also be introduced due to some other changes. And it would be nice if it wont take a month.

SimonVorgers commented:
At the moment the e2e test `Should disable free shipping filter` in `tests/e2e/cypress/e2e/storefront/listing/disabled-filter-combinations.cy.js` is failing.

It fails because it cant satisfy that the shipping filter is disabled. 

![Test product filters get disabled if a combination is not possible -- Should disable free shipping filter (failed)](https://user-images.githubusercontent.com/92690387/213174448-2f9ca0f2-db51-4deb-9abf-639ac8cd81f4.png)

SimonVorgers commented:
I will close this for now due to inactivity. Whenever you get around to fix the issue, feel free to reopen the PR.
